### PR TITLE
[wip] Early draft of support for metrics in tracing

### PR DESCRIPTION
**Pull request by bryangarza**
(This is a draft to get some early feedback. Please share any thoughts you have on this!)

## Motivation

TL;DR - It would be really useful to be able to publish service-level and application-level metrics via tracing.

Currently, we can generate some tokio-runtime metrics via tokio-metrics. However, we basically can only log these metrics like a regular log line and analyze them by hand.

The opentelemetry crate already supports publishing metrics. However, because tracing itself doesn't have native support for metrics, the tracing-opentelemetry crate only supports publishing traces.

Therefore, I'm trying to find what is a good way to add into tracing the support for service-level and application-level metrics.

**Note that I'm only talking about OTel in order to explain what an end-to-end solution could look like, but a different layer should be able to do the same for other spcs/vendors. Part of the goal here is not tying this to any given sink.**

## Goal

The eventual goal would be that we should be able to do something like

```rust
// (these are just rough ideas)
metric_counter!(MY_METRIC_NAME, foo = "bar", baz = "qux");
// or
metric!(MY_METRIC_NAME, unit = Gauge, 5,  foo = "bar", baz = "qux");
```

or even integrate metrics into the `#[instrument]` annotation (to enable for example, publishing metrics on each API request count of a service using Hyper/Axum/Tonic).

From there, a concrete implementation of the `Collect` trait would implement an additional `metric(...)` method, which is how (for example) the tracing-opentelemetry Layer could call the OTel Metrics SDK to publish metrics out (or log the metric to some other file, call a different vendor SDK, etc etc).

## Solution

Right now what I did is I duplicated the `event(...)` method from the Collect trait. The definition of `Metric` is also currently just a copy of the `Event` definition.

I'd also need to crate some `metric!(...)` macros to go along with this (I haven't written them yet).

**Is this approach something that could work?** Should I be instead sticking this method in a separate trait? I figure the plumbing is identical for a metric than for an event, so it makes sense to just essentially have the same code path.

## PS

There might be some code that should say `Metric` but says `Event`... I copy pasted stuff here and there and it might not all be entirely renamed correctly, especially comments. But I'm more looking for feedback on the approach itself. I will clean up the code later and add tests etc, once I have more of an idea of if this is the right approach or if this is what we might want to do.

## Glossary

service-level metrics = metrics such as request count, latency, tokio-runtime metrics... metrics that are not tied to the business logic

application-level logic = this is tied to the business logic, oftentimes it might be specific error metrics unique to your own service. Or, for example, keeping separate counters to track different code paths that you are entering into.

## Alternatives
- Instead of duplicating a lot of the Event code, create a `metric!` macro that desugars to an Event. This would require much fewer changes to the codebase and the difference between a Metric and Event is quite small anyways (the delta is: including the metric itself).

### CODE REVIEW

File: tracing-core/src/metric.rs [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+///   associated with an metric should be in the metric's fields rather than in
+///   the textual message, as the fields are more structured.
+///
+/// [span]: super::span
+/// [fields]: super::field
+#[derive(Debug)]
+pub struct Metric<'a> {
+    fields: &'a field::ValueSet<'a>,
+    metadata: &'static Metadata<'static>,
+    parent: Parent,
```
Comment (bryangarza) [inline tracing-core/src/metric.rs:26]: This would have to be changed of course. Like I mention in the body of my PR, the usual data we gather for Events might not all be relevant for metrics.

File: tracing-core/src/metric.rs [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    pub fn child_of(
+        parent: impl Into<Option<Id>>,
+        metadata: &'static Metadata<'static>,
+        fields: &'a field::ValueSet<'_>,
+    ) {
+        let metric = Self::new_child_of(parent, metadata, fields);
+        crate::dispatch::get_default(|current| {
+            current.metric(&metric);
+        });
+    }
```
Comment (bryangarza) [inline tracing-core/src/metric.rs:80]: We probably wouldn't need this for metrics... I think?

File: tracing-core/src/metric.rs [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    /// Returns the new metric's explicitly-specified parent, if there is one.
+    ///
+    /// Otherwise (if the new metric is a root or is a child of the current span),
+    /// returns `None`.
+    pub fn parent(&self) -> Option<&Id> {
+        match self.parent {
+            Parent::Explicit(ref p) => Some(p),
+            _ => None,
+        }
+    }
```
Comment (bryangarza) [inline tracing-core/src/metric.rs:127]: Probably also not needed

**Comment by bryangarza** [comment]
Talked with @carllerche offline, and we think it might make more sense to go with an alternative approach of creating a `metric!` macro that desugars to an event (mentioned at the bottom of my PR). Will close this an open a different PR once I have a new set of changes.
